### ISSUE TITLE
identify quality videos with classification method

### ISSUE THREAD

**Issue by avalanchesiqi** [avalanchesiqi/yt-engagement#2]
6 sets of predictors, but change to classification methods

**Comment by avalanchesiqi** [comment]
![relative_engagement_roc_explanation](https://user-images.githubusercontent.com/3666911/29699978-74206e36-89a4-11e7-9a6e-91d58e87184e.png)

One question remains in high engagement videos identification is that "why different thresholds exhibit the same trend". I did some investigation and found that resulted from a property of ROC, which is insensitive to class balance.

The figure above illustrates how ROC and PR curve react to different engagement thresholds. Here we are dealing with 2 thresholds, define as
_engagement threshold_: the minimal relative engagement so that we **label** it as a high engagement video, applied on observed data.
_varying thresholds_: the minimal relative engagement so that we **predict** it as a high engagement video, applied on predict data. This thresholds are necessary to construct ROC curve and PR curve.

I made illustrative examples on 2 predictors, "Content" predictor and "All" predictor.

- comparison between different predictors.
Consider the identification results (left 2 columns), we plot the prediction results with _engagement threshold_ as 0.5 and 0.9 respectively. The truly high engagement videos are colored as "red" while low engagement videos as "blue". X-axis is the probability of video being high engagement. We can see "All" predictors have a clear bimodal distribution, suggesting a better predictor to distinguish video engagement. "Content" predictor shows a negligible trend of bimodal.

- why ROC is insensitive to unbalanced class.
ROC curve is generated by true positive rate (TPR) and false positive rate (FPR) given _varying thresholds_. Consider when FPR equals to 0.5 on "All" predictor, which means 50% of negative instances are classified as positive (right 50% of blue histogram).

Note here how _engagement threshold_ works differently with _varying thresholds_. _engagement threshold_ only applies on observation and doesn't alter prediction probability.

When _engagement threshold_ equals to 0.5, we can calculate _varying threshold_ equals to 0.392 at this moment, and TPR equals to 0.912. However, when change _engagement threshold_ to 0.9, we get _varying threshold_ 0.489 and TPR 0.899. (Solid line in 2nd column). Here we can see very close (FPR, TPR) pairs are actually generated by different _varying threshold_ (circles in 3rd column). That's the reason why we observe similar trend for different _engagement threshold_. In other words, ROC curve is insensitive to class unbalance and will generate very close (but not identical) ROC curve. (3rd column)

- why PR-curve is more informative on unbalanced class.
Intuitively we can see 0.5 _engagement threshold_ separates bimodal better than 0.9 _engagement threshold_  but ROC curve fails to capture this discrepancy. PR curve is generated by precision and recall given _varying thresholds_. It uses exactly the same input (label, probability) as ROC curve. Let's consider a scenario where recall equals to 0.5, which means 50% of positive instances are classified as positive (right 50% of red histogram). We then calculate _varying threshold_ = 0.641 (dashed line in 2nd column), precision = 0.847 for _engagement threshold_ = 0.5, and _varying threshold_ = 0.795, precision = 0.447 for _engagement threshold_ = 0.9. _engagement threshold_ = 0.5 yields a much higher precision rate (circle in 4th columns), suggesting more accurate when recall rate is the same.

In conclusion, I think PR-curve is more suitable in our cases, and it captures the effect of skewed classes and depicts that the task gets harder to identify higher engagement videos.

*note: here I still use the result of regressor, since it's a value ranging from 0 to 1, so I consider it as probability in classification task.
